compiler/src/iree/compiler/Codegen/Common/LinkTuningSpecsPass.cpp

```diff
@@ -61,6 +62,178 @@ static bool consumesInputOp(NamedSequenceOp op) {
+    unsigned suffix;
+    if (!specName.substr(pos + 1).getAsInteger(10, suffix)) {
+      return potentialBase.str(); // If valid number, return base name
+    }
+  }
+  return specName.str(); // Otherwise, return full name
+}
+
+static std::string
+getUniqueSpecName(StringRef specName,
+                  llvm::StringMap<unsigned> &specNameCounts) {
+  std::string specBaseName = getBaseName(specName);
+  unsigned specNameSeenCount = specNameCounts[specBaseName]++;
+
+  if (specNameSeenCount == 0)
+    return specBaseName;
+
+  return llvm::formatv("{}_{}", specBaseName, specNameSeenCount).str();
+}
+
+struct TuningSpecsToMerge {
+  SmallVector<NamedSequenceOp> namedSequenceOpsToMove;
+  llvm::DenseMap<NamedSequenceOp, ForeachMatchOp> namedSequenceToForeachMatch;
+};
+
+// Updates the mapping of `NamedSequenceOp` to `ForeachMatchOp`
+// by processing matchers and actions inside the given `ForeachMatchOp`.
+static void
+updateForeachMatchMappings(ForeachMatchOp foreachMatch, ModuleOp module,
+                           llvm::DenseMap<NamedSequenceOp, ForeachMatchOp>
+                               &namedSequenceToForeachMatch) {
+  // Process matchers.
+  for (auto matcher : foreachMatch.getMatchers()) {
+    if (auto matcherSymRef = mlir::dyn_cast<mlir::SymbolRefAttr>(matcher)) {
+      if (auto matcherOp =
+              SymbolTable::lookupNearestSymbolFrom<NamedSequenceOp>(
+                  module, matcherSymRef)) {
+        namedSequenceToForeachMatch[matcherOp] = foreachMatch;
+      }
+    }
+  }
+
+  // Process actions.
+  for (auto action : foreachMatch.getActions()) {
+    if (auto actionSymRef = mlir::dyn_cast<mlir::SymbolRefAttr>(action)) {
+      if (auto actionOp = SymbolTable::lookupNearestSymbolFrom<NamedSequenceOp>(
+              module, actionSymRef)) {
+        namedSequenceToForeachMatch[actionOp] = foreachMatch;
+      }
+    }
+  }
+}
+
+// Collects tuning specs (`NamedSequenceOp`s) from all inner modules within the
+// given input model, and organizes them into a `TuningSpecsToMerge` struct.
+// - `namedSequenceOpsToMove`: Contains `NamedSequenceOp`s that either:
+//     - Are not named `__kernel_config`.
+//     - Do not have the `iree_codegen.tuning_spec_entrypoint` attribute.
+// - `namedSequenceToForeachMatch`: Maps `NamedSequenceOp`s used as matchers or
+//    actions to their corresponding `ForeachMatchOp`.
+static TuningSpecsToMerge collectTuningSpecsToMerge(ModuleOp module) {
+  TuningSpecsToMerge tuningSpecs;
+  for (auto innerModule : module.getBody()->getOps<ModuleOp>()) {
+    for (auto namedSequenceOp :
+         innerModule.getBody()->getOps<NamedSequenceOp>()) {
+      if (namedSequenceOp.getSymName() != kKernelConfigSpecName ||
+          !namedSequenceOp->hasAttr(kTuningSpecEntrypointAttrName)) {
+        tuningSpecs.namedSequenceOpsToMove.push_back(namedSequenceOp);
+        continue;
+      }
+
+      // Extract the single ForeachMatchOp inside namedSequenceOp.
+      ForeachMatchOp foreachMatch =
```

Comment:
```suggestion
      auto foreachMatch =
```
See https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable